crates/ruff_linter/src/rules/flake8_bugbear/rules/loop_iterator_mutated.rs

```diff
@@ -0,0 +1,185 @@
+use ruff_diagnostics::Violation;
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{
+    visitor::{self, Visitor},
+    Expr, ExprAttribute, ExprCall, ExprContext, ExprName, ExprSubscript, Stmt, StmtDelete, StmtFor,
+};
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+use ruff_diagnostics::Diagnostic;
+
+static MUTATING_FUNCTIONS: &'static [&'static str] = &[
+    "append", "sort", "reverse", "remove", "clear", "extend", "insert", "pop", "popitem",
+];
+
+/// ## What it does
+/// Checks for mutation of the iterator of a loop in the loop's body
+///
+/// ## Why is this bad?
+/// Changing the structure that is being iterated over will usually lead to
+/// unintended behavior as not all elements will be addressed.
+///
+/// ## Example
+/// ```python
+/// some_list = [1,2,3]
+/// for i in some_list:
+///   some_list.remove(i) # this will lead to not all elements being printed
+///   print(i)
+/// ```
+///
+///
+/// ## References
+/// - [Python documentation: Mutable Sequence Types](https://docs.python.org/3/library/stdtypes.html#typesseq-mutable)
+#[violation]
+pub struct LoopIteratorMutated;
+
+impl Violation for LoopIteratorMutated {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("editing a loop's mutable iterable often leads to unexpected results/bugs")
+    }
+}
+
+fn _to_name_str(node: &Expr) -> String {
+    match node {
+        Expr::Name(ExprName { id, .. }) => {
+            return id.to_string();
+        }
+        Expr::Attribute(ExprAttribute {
+            range: _,
+            value,
+            attr,
+            ..
+        }) => {
+            let mut inner = _to_name_str(value);
+            match inner.as_str() {
+                "" => {
+                    return "".into();
+                }
+                _ => {
+                    inner.push_str(".");
+                    inner.push_str(attr);
+                    return inner;
+                }
+            }
+        }
+        Expr::Call(ExprCall { range: _, func, .. }) => {
+            return _to_name_str(func);
+        }
+        _ => {
+            return "".into();
+        }
+    }
+}
+// B038
+pub(crate) fn loop_iterator_mutated(checker: &mut Checker, stmt_for: &StmtFor) {
+    let StmtFor {
+        target: _,
+        iter,
+        body,
+        orelse: _,
+        is_async: _,
+        range: _,
+    } = stmt_for;
+    let name;
+
+    match iter.as_ref() {
+        Expr::Name(ExprName { .. }) => {
+            name = _to_name_str(iter.as_ref());
+        }
+        Expr::Attribute(ExprAttribute { .. }) => {
+            name = _to_name_str(iter.as_ref());
+        }
+        _ => {
+            println!("Shouldn't happen");
```

Comment:
We should remove this -- if it's an actual invariant, we should find a way to enforce it with the type system! And if it's actually an error, we should log it with `error!` or similar.

Reply:
How wold you go about ensuring this with the type system?

Reply:
I think you should use the [`unreachable!()`](https://doc.rust-lang.org/std/macro.unreachable.html) macro.